core/src/main/java/io/neba/core/resourcemodels/mapping/CustomFieldMappers.java

```diff
@@ -0,0 +1,114 @@
+ */
+@Service
+public class CustomFieldMappers {
+    /**
+     * @author Olaf Otto
+     */
+    public static class AnnotationMapping {
+        private final Annotation annotation;
+        private final AnnotatedFieldMapper mapper;
+
+        public AnnotationMapping(Annotation annotation, AnnotatedFieldMapper mapper) {
+            this.annotation = annotation;
+            this.mapper = mapper;
+        }
+
+        public Annotation getAnnotation() {
+            return annotation;
+        }
+
+        public AnnotatedFieldMapper getMapper() {
+            return mapper;
+        }
+    }
+
+    private static final Collection<AnnotationMapping> NULL = emptyList();
+    private final ConcurrentDistinctMultiValueMap<Field, AnnotationMapping> cache
+          = new ConcurrentDistinctMultiValueMap<Field, AnnotationMapping>();
+    private final ConcurrentDistinctMultiValueMap<Class<? extends Annotation>, AnnotatedFieldMapper> fieldMappers
+          = new ConcurrentDistinctMultiValueMap<Class<? extends Annotation>, AnnotatedFieldMapper>();
+
+    public synchronized void add(AnnotatedFieldMapper<?, ?> mapper) {
+        if (mapper == null) {
+            throw new IllegalArgumentException("Method argument mapper must not be null.");
+        }
+        this.fieldMappers.put(mapper.getAnnotationType(), mapper);
+        this.cache.clear();
+    }
+
+    public synchronized void remove(AnnotatedFieldMapper<?, ?> mapper) {
+        if (mapper == null) {
+            throw new IllegalArgumentException("Method argument mapper must not be null.");
+        }
+        this.fieldMappers.remove(mapper.getAnnotationType());
+        this.cache.clear();
+    }
+
+    public Collection<AnnotationMapping> get(MappedFieldMetaData metaData) {
+        if (metaData == null) {
+            throw new IllegalArgumentException("Method argument metaData must not be null.");
+        }
+
+        Collection<AnnotationMapping> mappers = this.cache.get(metaData.getField());
+
+        if (mappers == NULL) {
+            return NULL;
+        }
+
+        if (mappers != null) {
+            return mappers;
+        }
+
+        List<AnnotationMapping> compatibleMappers = new ArrayList<AnnotationMapping>();
+        for (Annotation annotation : metaData.getAnnotations()) {
+            Collection<AnnotatedFieldMapper> mappersForAnnotation = this.fieldMappers.get(annotation.annotationType());
+            if (mappersForAnnotation == null) {
+                continue; // with next element
+            }
+            for (AnnotatedFieldMapper<?, ?> mapper:  mappersForAnnotation) {
+                if (mapper.getFieldType().isAssignableFrom(metaData.getType())) {
+                    compatibleMappers.add(new AnnotationMapping(annotation, mapper));
+                }
+            }
+        }
+        mappers = compatibleMappers.isEmpty() ? NULL : compatibleMappers;
+
+        synchronized (this) {
+            this.cache.put(metaData.getField(), mappers);
```

Comment:
I don't know the implementation of ConcurrentDistinctMultiValueMap, but according to the name I would expect that this does not have to be synchronized here again.